server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -423,6 +428,12 @@ private void handleJoinRequest(JoinRequest joinRequest, JoinHelper.JoinCallback
         assert Thread.holdsLock(mutex) == false;
         assert getLocalNode().isMasterNode() : getLocalNode() + " received a join but is not master-eligible";
         logger.trace("handleJoinRequest: as {}, handling {}", mode, joinRequest);
+
+        if (singleNodeDiscovery && joinRequest.getSourceNode().equals(getLocalNode()) == false) {
+            joinCallback.onFailure(new IllegalStateException("cannot join node with single-node discovery"));
```

Comment:
Suggest `cannot join node with [discovery.type] set to [single-node]` so that it's clearer what setting to look for.